### PR TITLE
#2043 High Availability (HA) Subsystem Improvement 

### LINKED ISSUE #2043: High Availability (HA) Subsystem Improvement Plan
# ArcadeDB High Availability (HA) Subsystem Improvement Plan

## Executive Summary

This document outlines a comprehensive plan to improve the ArcadeDB High Availability subsystem based on analysis of the `feature/2043-ha-test` branch compared to `main`. The branch introduces several key enhancements:

1. **ServerInfo Record** - New structured representation for server addresses with alias support
2. **HACluster Class** - Centralized cluster configuration management
3. **Docker/K8s Discovery** - Alias mechanism for dynamic container environments
4. **Resilience Module** - Testcontainers + Toxiproxy-based fault injection testing

However, several issues and incomplete implementations need attention before merging.

---

### Changes from Main Branch

**Total Changes**: 67 files, +1,584 lines, -653 lines

**Key Architectural Changes**:

| Component | Main Branch | Feature Branch |
|-----------|-------------|----------------|
| Server Identifier | `String` (host:port) | `ServerInfo` record (host, port, alias) |
| Cluster Configuration | `Set<String>` | `HACluster` class |
| Replica Connections | `Map<String, L2RNetworkExecutor>` | `Map<ServerInfo, L2RNetworkExecutor>` |
| Server Address | `String serverAddress` | `ServerInfo serverAddress` |
| Enum Naming | `SCREAMING_CASE` | `PascalCase` (Java standard) |

### New Components

1. **`resilience` Maven Module** - Dedicated module for network failure testing
2. **`ContainersTestTemplate`** - Base class for Testcontainers-based HA tests
3. **`ServerInfo` Record** - Type-safe server address representation
4. **`HACluster` Class** - Cluster topology abstraction

---

#### 1. Incomplete Alias Resolution in Server Discovery
**Location**: `HAServer.java:1062`, `SimpleHaScenarioIT.java:29-30`

The alias mechanism `{arcade2}proxy:8667` is parsed but not fully resolved during cluster formation:

```java
// Error seen in logs:
// Error connecting to the remote Leader server {proxy}proxy:8666
// (error=Invalid host proxy:8667{arcade3}proxy:8668)
```

**Root Cause**: When building the server list, aliases are being concatenated incorrectly.

#### 2. `setServerAddresses()` Method is Commented Out
**Location**: `HAServer.java:540-560`

The method that should update cluster configuration from replicas is entirely commented out, leaving dead code.

#### 3. `removeServer()` Still Uses String
**Location**: `HAServer.java:749`

```java
public void removeServer(final String remoteServerName) {
    final Leader2ReplicaNetworkExecutor c = replicaConnections.remove(remoteServerName);
```

This is incompatible with the new `Map<ServerInfo, ...>` signature and will cause ClassCastException.

#### 4. HTTP Address Propagation Disabled
**Location**: `HAServer.java:697-728`

Methods `setReplicasHTTPAddresses()` and `getReplicaServersHTTPAddressesList()` are commented out, breaking client redirect functionality.

#### 5. Test Assertions Wrong in ThreeInstancesScenarioIT
**Location**: `ThreeInstancesScenarioIT.java:103-105`

```java
// When arcade1 is disconnected, arcade2 and arcade3 should have data
// But test asserts on arcade1 which is disconnected!
db1.assertThatUserCountIs(130);  // arcade1 is DISCONNECTED - can't assert
db2.assertThatUserCountIs(130);  // correct
db3.assertThatUserCountIs(130);  // correct
```

#### 6. ReplicationServerQuorumNoneIT Reliability
**Location**: `ReplicationServerQuorumNoneIT.java:40-47`

While timeout increases help, the fundamental issue is that async replication with `QUORUM=NONE` allows unbounded queue growth. The test reduces load but doesn't address the underlying design issue.

#### 7. Missing ServerInfo equals/hashCode Consideration
**Location**: `HAServer.java:83-97`

The `ServerInfo` record should have explicit documentation about identity semantics. Currently it compares all three fields, but for cluster membership, only `host:port` should matter.

#### 8. Thread Safety Concerns
**Location**: `HAServer.java:82` (cluster field)

```java
private HACluster cluster;  // Not volatile, accessed from multiple threads
```

The `cluster` field is modified in `setServerAddresses()` (when it works) and read during elections without synchronization.

#### 9. Debug Logging Left in Code
**Location**: `HAServer.java:640, 1078, 1085`

Several `Level.INFO` log statements should be `Level.FINE`:
```java
LogManager.instance().log(this, Level.INFO, "Sending request (%s) to %s", ...);
LogManager.instance().log(this, Level.INFO, "Creating client connection to '%s'", ...);
```

#### 10. Commented-Out Code in ReplicationServerIT
**Location**: `ReplicationServerIT.java:1` - File starts with `ver /*` (typo)

---

### PR DESCRIPTION
**Pull request by robfrank**
This pull request addresses several critical issues and improvements in the High Availability (HA) subsystem, focusing on bug fixes, protocol consistency, and test reliability. The main changes include fixing alias resolution for cluster discovery, correcting type mismatches in server removal, re-enabling HTTP address propagation for client redirection, and fixing test assertions in resilience scenarios. Additionally, the CI workflow is updated to properly handle resilience tests.

**HA Subsystem Bug Fixes and Improvements:**

* Alias Resolution
  - Implemented a `resolveAlias()` method in `HAServer.java` to ensure that server addresses containing aliases (e.g., `{arcade2}proxy:8667`) are correctly resolved to actual host:port values during leader connection, fixing cluster formation issues in Docker/K8s environments.
  - Added comprehensive tests for alias resolution, including edge cases and multiple alias scenarios.

* Server Removal Consistency
  - Updated the `removeServer()` method in `HAServer.java` to accept `ServerInfo` objects instead of `String`, aligning with the migration to `ServerInfo`-based replica connection maps. A backward-compatible method accepting `String` was also added.
  - Added a helper method `findByHostAndPort()` to `HACluster` for easier `ServerInfo` lookups and fixed enum naming conventions for consistency.

* HTTP Address Propagation
  - Re-enabled and updated the mechanism for propagating HTTP addresses from replicas to the leader, allowing clients to be redirected to the correct replica HTTP endpoints. This includes protocol changes where replicas send their HTTP address to the leader during connection, and the leader tracks these addresses in a map.
  - Updated `GetServerHandler` to return the correct HTTP addresses for client redirection and ensured cleanup of HTTP address mappings when servers are removed.

**Test Reliability and Workflow:**

* Test Assertion Correction
  - Fixed incorrect assertions in the `ThreeInstancesScenarioIT` test by removing assertions on a disconnected node (arcade1) and clarifying comments to reflect the actual test scenario, ensuring the test only checks nodes that are connected and can replicate data.

* CI Workflow Update
  - Modified the GitHub Actions workflow to exclude the `resilience` module from the main integration test job and introduced a dedicated `java-resilience-tests` job to run resilience-specific tests separately. [[1]](diffhunk://#diff-de8e1c99a0afd1ff3311ccc77eb1d3495d898aea420ba3450bf6edb8794aa660L182-R182) [[2]](diffhunk://#diff-de8e1c99a0afd1ff3311ccc77eb1d3495d898aea420ba3450bf6edb8794aa660R293-R336)

These changes collectively improve the robustness, correctness, and maintainability of the HA subsystem and associated tests.

## Related issues

#2043 

## Checklist

- [x] I have run the build using `mvn clean package` command
- [x] My unit tests cover both failure and success scenarios

### CODE REVIEW

File: network/src/main/java/com/arcadedb/remote/RemoteDatabase.java [hunk 1]
```diff
@@ -400,7 +400,7 @@ public ResultSet command(final String language, final String command, final Cont
     checkDatabaseIsOpen();
     stats.commands.incrementAndGet();
 
-    return (ResultSet) databaseCommand("command", language, command, params, true,
```
Comment (lvca) [inline network/src/main/java/com/arcadedb/remote/RemoteDatabase.java:476]: Why are you setting false = not require a leader? This means the operation will go to any server, but if it's an update (query is idempotent) will bounce to the leader in 2 passes

**Review by lvca — APPROVED** [review]

**Comment by robfrank** [comment]
Closing. A new HA stack based on Apache Ratis is now in plae
